### PR TITLE
Add static code analysis with PHPStan level 0

### PR DESCRIPTION
**Pull request by Art4**
This PR adds [PHPStan](https://phpstan.org/) as static code analysis tool and adds a new composer script `phpstan`. Using this script PHPStan on [level 0](https://phpstan.org/user-guide/rule-levels) founds 67 errors, that will be fixed in this PR.

<details><summary>Details</summary>

```bash

 $ bin/composer.phar run phpstan
Xdebug: [Step Debug] Could not connect to debugging client. Tried: localhost:9000 (through xdebug.client_host/xdebug.client_port).
> phpstan analyze --memory-limit 1024M --configuration .phpstan.neon
Xdebug: [Step Debug] Could not connect to debugging client. Tried: localhost:9000 (through xdebug.client_host/xdebug.client_port).
 1014/1014 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ------------------------------------------------------------------------------ 
  Line   App/Mode.php                                                                  
 ------ ------------------------------------------------------------------------------ 
  :213   Class Friendica\App\Mode referenced with incorrect case: Friendica\App\MODE.  
 ------ ------------------------------------------------------------------------------ 

 ------ ----------------------------------------------------------------------------------- 
  Line   BaseCollection.php                                                                 
 ------ ----------------------------------------------------------------------------------- 
  :92    Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :105   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :115   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :131   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
 ------ ----------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------ 
  Line   Console/Addon.php                                                                               
 ------ ------------------------------------------------------------------------------------------------ 
  :146   Method Friendica\Console\Addon::list() should return bool|int but return statement is missing.  
 ------ ------------------------------------------------------------------------------------------------ 

 ------ --------------------------------------------------------------------- 
  Line   Console/AutomaticInstallation.php                                    
 ------ --------------------------------------------------------------------- 
  :137   Instantiated class Friendica\Console\Exception not found.            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols  
 ------ --------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------- 
  Line   Console/Contact.php                                                                                 
 ------ ---------------------------------------------------------------------------------------------------- 
  :159   Method Friendica\Console\Contact::addContact() should return bool but return statement is missing.  
 ------ ---------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------- 
  Line   Console/FixAPDeliveryWorkerTaskParameters.php                                                
 ------ --------------------------------------------------------------------------------------------- 
  :71    Access to an undefined property Friendica\Console\FixAPDeliveryWorkerTaskParameters::$l10n.  
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property         
 ------ --------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------- 
  Line   Console/User.php                                                                                 
 ------ ------------------------------------------------------------------------------------------------- 
  :461   Method Friendica\Console\User::configUser() should return bool but return statement is missing.  
 ------ ------------------------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------- 
  Line   Content/Post/Entity/PostMedia.php                                                  
 ------ ----------------------------------------------------------------------------------- 
  :231   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :258   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
 ------ ----------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------- 
  Line   Content/Widget/ContactBlock.php                                                                                    
 ------ ------------------------------------------------------------------------------------------------------------------- 
  :33    Template type widget of method Friendica\Content\Widget\ContactBlock::getHTML() is not referenced in a parameter.  
 ------ ------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------ 
  Line   Content/Widget/VCard.php                                                                                    
 ------ ------------------------------------------------------------------------------------------------------------ 
  :35    Template type widget of method Friendica\Content\Widget\VCard::getHTML() is not referenced in a parameter.  
 ------ ------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------ 
  Line   Core/Cache/Type/APCuCache.php                                     
 ------ ------------------------------------------------------------------ 
  :47    Class APCuIterator referenced with incorrect case: APCUIterator.  
  :128   Class APCuIterator referenced with incorrect case: APCUIterator.  
 ------ ------------------------------------------------------------------ 

 ------- ------------------------------------------------------ 
  Line    Model/Contact.php                                     
 ------- ------------------------------------------------------ 
  :2811   Cannot unset offset 'avatar' on array{photo: mixed}.  
 ------- ------------------------------------------------------ 

 ------- -------------------------- 
  Line    Model/Item.php            
 ------- -------------------------- 
  :3204   Undefined variable: $set  
 ------- -------------------------- 

 ------ -------------------------------- 
  Line   Module/Api/Mastodon/Search.php  
 ------ -------------------------------- 
  :161   Undefined variable: $since_id   
 ------ -------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   Module/Debug/ActivityPubConversion.php                                                                                                                            
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  :26    Inner named functions are not supported by PHPStan. Consider refactoring to an anonymous function, class method, or a top-level-defined function. See issue #165  
         (https://github.com/phpstan/phpstan/issues/165) for more details.                                                                                                 
  :46    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :50    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :58    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :96    Undefined variable: $body                                                                                                                                         
  :110   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :117   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   Module/Debug/Babel.php                                                                                                                                            
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  :32    Inner named functions are not supported by PHPStan. Consider refactoring to an anonymous function, class method, or a top-level-defined function. See issue #165  
         (https://github.com/phpstan/phpstan/issues/165) for more details.                                                                                                 
  :45    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :51    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :57    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :62    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :73    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :79    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :85    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :95    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :101   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :109   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :113   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :119   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :124   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :135   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :144   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :150   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :161   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :168   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :180   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :185   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :196   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :213   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :219   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :225   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :231   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :242   Function twitter_statuses_show not found.                                                                                                                         
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :249   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :266   Function twitter_media_entities not found.                                                                                                                        
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :270   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :273   Function twitter_expand_entities not found.                                                                                                                       
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :277   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :282   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ --------------------------------------------------------------------------------------------------------- 
  Line   Module/Debug/ItemBody.php                                                                                
 ------ --------------------------------------------------------------------------------------------------------- 
  :35    Method Friendica\Module\Debug\ItemBody::content() should return string but return statement is missing.  
 ------ --------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------- 
  Line   Object/Api/Mastodon/Preferences.php                                                                      
 ------ --------------------------------------------------------------------------------------------------------- 
  :42    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$posting:default:visibility.  
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :43    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$posting:default:sensitive.   
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :44    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$posting:default:language.    
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :45    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$reading:expand:media.        
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :46    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$reading:expand:spoilers.     
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
 ------ --------------------------------------------------------------------------------------------------------- 

 ------- ------------------------------------------------ 
  Line    Protocol/ActivityPub/Receiver.php               
 ------- ------------------------------------------------ 
  :1900   Cannot unset offset -1 on array<mixed, mixed>.  
 ------- ------------------------------------------------ 

 ------ --------------------------------------------------------------------------------------------------------- 
  Line   Util/Temporal.php                                                                                        
 ------ --------------------------------------------------------------------------------------------------------- 
  :347   Method Friendica\Util\Temporal::getRelativeDate() should return string but return statement is missing.  
 ------ --------------------------------------------------------------------------------------------------------- 

                                                                                                                        
 [ERROR] Found 67 errors                                                                                                
                                                                                                                        

Script phpstan analyze --memory-limit 1024M --configuration .phpstan.neon handling the phpstan event returned with error code 1
```
</details> 

Requires merge of #14526, #14539 and #14523.

### CODE REVIEW

**Comment by MrPetovan** [comment]
It looks like this PR also includes https://github.com/friendica/friendica/pull/14523, is this intended?

**Comment by Art4** [comment]
> It looks like this PR also includes #14523, is this intended?

Yes, this is intended because this PR installs PHPStan using composer. If I do this with composer v1 I will have to resolve merge conflict in the `composer.lock` later.

Another reason is that I don't have to wait for the previous PRs to merge but I can already work with their improvements. This PR will stay as a draft as long the required PRs are merged. If they are rejected I will have to revert their changes from this PR.

**Comment by Art4** [comment]
This PR is ready for review. Because this PR requires #14526 and #14523 one can compare the changes directly between the branches here: https://github.com/Art4/friendica/compare/improve-tests...Art4:friendica:add-static-code-analysis?expand=1

I'm not sure why the CI pipelines fails because looking at the details for the specific pipelines I only see a grey page.

**Comment by Art4** [comment]
While we are waiting for #14523 one can look at the relevant changes comparing the branches:

https://github.com/Art4/friendica/compare/improve-tests...Art4:friendica:add-static-code-analysis

**Comment by MrPetovan** [comment]
> I'm not sure why the CI pipelines fails because looking at the details for the specific pipelines I only see a grey page.

Same here, I can see a couple Javascript errors in https://ci.friendi.ca/repos/2/pipeline/2448/1 but the JS file is minified so it is impossible to figure out the actual cause.

**Comment by annando** [comment]
Yeah, the pipeline currently is a huge problem. Possibly as some interim solution we should activate an additional pipeline on Github (like @Art4 already suggested) that does the same job. I just don't know how to do it.

**Comment by Art4** [comment]
I'm already trying to manually start github actions, but there is an error with mariadb.

https://github.com/Art4/friendica/actions/runs/11724292576/job/32657882187

Branch: https://github.com/Art4/friendica/tree/restore-github-actions

I will create a PR where we can investigate this issue.

**Comment by annando** [comment]
@Art4 the pipeline now works again.

**Comment by Art4** [comment]
I fixed the ci pipelines. After #14526 is merged this PR can be reviewed.

**Review by MrPetovan — COMMENTED** [review]
Don't forget to rebase your branch now that the test PR was merged!

File: src/BaseCollection.php
Comment (MrPetovan) [inline src/BaseCollection.php]: This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

File: src/BaseCollection.php
Comment (MrPetovan) [inline src/BaseCollection.php]: This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

File: src/BaseCollection.php
Comment (MrPetovan) [inline src/BaseCollection.php]: This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

File: src/BaseCollection.php
Comment (MrPetovan) [inline src/BaseCollection.php]: This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

File: src/Model/Item.php
Comment (MrPetovan) [inline src/Model/Item.php:3201]: Good catch!

File: src/Module/Debug/ActivityPubConversion.php
Comment (MrPetovan) [inline src/Module/Debug/ActivityPubConversion.php]: This used to be a quick helper function used only in Debug classes. If you must make it into a full-fledged class method, please move it to the `Content\Text\HTML` class and rename it `visibleWhitespace`.

File: src/Module/Debug/ItemBody.php
Comment (MrPetovan) [inline src/Module/Debug/ItemBody.php:14]: I've used this convention extensively through the code base where I only import the HTTP exceptions once, otherwise I feel like it clutters the `use` block for little benefit.

File: src/Protocol/ActivityPub/Receiver.php
Comment (MrPetovan) [inline src/Protocol/ActivityPub/Receiver.php]: What's the rationale behind this? I thought `unset` was permissive when it comes to undefined symbols including no-existent array keys?

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/BaseCollection.php
Comment (Art4) [reply src/BaseCollection.php]: You are right. I wasn't aware that this class would also be extended. The problem with using `static` is that the extending class can overwrite the `__construct()` method with different parameters. See https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static

I will mute this error for now by using `get_class($this);` but this has to be refactored in the future.

File: src/Module/Debug/ItemBody.php
Comment (Art4) [reply src/Module/Debug/ItemBody.php:14]: Cluttering the `use` block is a good thing because it shows you the code smell. If the `use` block is to big its an indicator that the class maybe does too much.

File: src/Protocol/ActivityPub/Receiver.php
Comment (Art4) [reply src/Protocol/ActivityPub/Receiver.php]: It makes us more aware of the array values. Yes, `unset()` is permissive for undefined symbols but we should move to more specific types. Strict types in arrays will even let PHPStan recognize in higher levels when the call of `unset()` is unnecessary.

I improved the code a little bit so we only have one check and no unnecessary calls of `unset()`.

File: src/Module/Debug/ActivityPubConversion.php
Comment (Art4) [reply src/Module/Debug/ActivityPubConversion.php]: The problem here was the definition of a named function inside a method. This way you can't call the method twice, see https://3v4l.org/Y7iu7

If this quick helper function is only for debugging I would keep it as a private method. There is no problem with using a "full-fledged" method for this. I would not recommend to create a new dependency to the `Content\Text\HTML`, because this will extend our API surface that we will have to maintain. If you like I can also create an anonymous function inside the method to fix this.

File: src/BaseCollection.php
Comment (Art4) [reply src/BaseCollection.php]: fixed

File: src/BaseCollection.php
Comment (Art4) [reply src/BaseCollection.php]: fixed.

File: src/BaseCollection.php
Comment (Art4) [reply src/BaseCollection.php]: fixed

File: src/BaseCollection.php
Comment (MrPetovan) [reply src/BaseCollection.php]: I saw this, and in one instance we are extending the constructor, so we can't use any of the remediation solutions. It's another one of these "it might break if" issues that currently don't apply to us that we unfortunately have to mute rather than "fix".

File: src/Module/Debug/ActivityPubConversion.php
Comment (MrPetovan) [reply src/Module/Debug/ActivityPubConversion.php]: Interesting, I didn't consider that. However the `rawContent()`, `content()` and `post()` methods are meant to be only called once per request so the issue never arose. Was it stunting the tests by any chance?

As for the dependency to `Content\Text\HTML`, it's the one area where I know how to write extensive unit tests since methods are pure and don't depend on any other service. This makes me confident about using it as a dependency itself.

If it still had been used only in one place, I would have preferred going the private function route, but since it's now used in two different places, I'd rather it being centralized in a single place which I think should be `Content\Text\HTML`.

File: src/Module/Debug/ItemBody.php
Comment (MrPetovan) [reply src/Module/Debug/ItemBody.php:14]: In general I'd agree, however I think exceptions are the exception. For best results, exceptions should be as unique as possible (although this can also be achieved with error codes). We already are simplifying to only HTTP-like exceptions, but there still are dozens of them, and I don't think using several of them where appropriate is a code smell.

File: src/Protocol/ActivityPub/Receiver.php
Comment (MrPetovan) [inline src/Protocol/ActivityPub/Receiver.php]: Please align the assignment operators.
```suggestion
		$object_data['unlisted']       = false;
```

File: src/BaseCollection.php
Comment (MrPetovan) [reply src/BaseCollection.php]: I'm also curious about the different between `static` and `get_class($this)` in this case, why does it mute this warning?

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/BaseCollection.php
Comment (Art4) [reply src/BaseCollection.php]: > I'm also curious about the different between `static` and `get_class($this)` in this case, why does it mute this warning?

I cannot say for sure but at level 3 PHPStan will start again complaining about this code.

File: src/Module/Debug/ActivityPubConversion.php
Comment (Art4) [reply src/Module/Debug/ActivityPubConversion.php]: > Was it stunting the tests by any chance?

I'm not sure that this method is called in tests but if it was called twice you definitely would see it.

> As for the dependency to `Content\Text\HTML`, it's the one area where I know how to write extensive unit tests since methods are pure and don't depend on any other service. This makes me confident about using it as a dependency itself.
>
> If it still had been used only in one place, I would have preferred going the private function route, but since it's now used in two different places, I'd rather it being centralized in a single place which I think should be `Content\Text\HTML`.

It is not about the `Content\Text\HTML` class but about the `ActivityPubConversion` class that worries me more. In my opinion this small html code would be best placed in the `view/templates/debug/activitypubconversion.tpl` template file. This should be done in a separate PR.

I will change this code to an anonymous function, do you agree?

File: src/Module/Debug/ItemBody.php
Comment (Art4) [reply src/Module/Debug/ItemBody.php:14]: Fully agreed. In this class imo there aren't much dependencies. But with every new used class one will get a better feeling if the class starts to getting a monolith.

Another great thing about writing every FQCN in the `use` block is that you can easily find all places where a class is used i.e. searching for `use Friendica\Network\HTTPException\NotFoundException`.

File: src/Module/Debug/ActivityPubConversion.php
Comment (MrPetovan) [reply src/Module/Debug/ActivityPubConversion.php]: Absolutely, it isn't big enough to warrant a proper function/method.

File: src/Module/Debug/ItemBody.php
Comment (MrPetovan) [reply src/Module/Debug/ItemBody.php:14]: I see what you mean, although it's only useful for full-text search, as static analyzers within IDEs will correctly identify usage even with the relative `use` calls.

File: src/Protocol/ActivityPub/Receiver.php
Comment (MrPetovan) [inline src/Protocol/ActivityPub/Receiver.php]: I thought you simplified to a single condition on line 1899, why the change of heart?

Comment (Art4) [reply src/Protocol/ActivityPub/Receiver.php]: Sorry, totally forgot about this: I updated to [PHPStan 2.0](https://phpstan.org/blog/phpstan-2-0-released-level-10-elephpants) and now PHPStan started complaining about this line too. :see_no_evil: 

I'll try to refactor the method again.

Comment (Art4) [reply src/Protocol/ActivityPub/Receiver.php]: I've refactored the whole method to check for `-1` as key upfront. This way a value with this key will never been set and don't has to be `unset()` again. 

Having some tests for this method would be nice but It would be to much work atm to mock the other private static methods.

**Review by MrPetovan — APPROVED** [review]

**Comment by MrPetovan** [comment]
Waiting for release to merge if I understood correctly.

**Comment by Art4** [comment]
> Waiting for release to merge if I understood correctly.

@MrPetovan A release should not slow down the development. Please see #14555.

**Pull request merged**
